src/rum/injection.cpp

```diff
@@ -0,0 +1,257 @@
+InjectionHandler::~InjectionHandler() {
+  if (injector_ != nullptr) {
+    injector_cleanup(injector_);
+  }
+}
+
+ngx_int_t InjectionHandler::on_rewrite_handler(ngx_http_request_t *r) {
+  ngx_table_elt_t *h =
+      static_cast<ngx_table_elt_t *>(ngx_list_push(&r->headers_in.headers));
+  if (h == nullptr) {
+    ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                  "RUM SDK injection failed: unable to add "
+                  "x-datadog-rum-injection-pending HTTP header");
+    return NGX_ERROR;
+  }
+
+  ngx_str_set(&h->key, "x-datadog-rum-injection-pending");
+  ngx_str_set(&h->value, "1");
+  h->lowcase_key = h->key.data;
+  h->hash = 1;
+
+  return NGX_DECLINED;
+}
+
+ngx_int_t InjectionHandler::on_header_filter(
+    ngx_http_request_t *r, datadog_loc_conf_t *cfg,
+    ngx_http_output_header_filter_pt &next_header_filter) {
+  assert(cfg->rum_snippet != nullptr);
+
+  if (!cfg->rum_enable) {
+    return next_header_filter(r);
+  }
+
+  if (auto injected_header = search_header(r, "x-datadog-rum-injected");
+      injected_header != nullptr) {
+    if (nginx::to_string_view(injected_header->value) == "1") {
+      ngx_log_error(NGX_LOG_INFO, r->connection->log, 0,
+                    "RUM SDK injection skipped: resource may already have RUM "
+                    "SDK injected.");
+      return next_header_filter(r);
+    }
+  }
+
+  if (r->header_only || r->headers_out.content_length_n == 0) {
+    ngx_log_error(NGX_LOG_INFO, r->connection->log, 0,
+                  "RUM SDK injection skipped: empty content");
+    return next_header_filter(r);
+  }
+
+  if (!is_html_content(&r->headers_out.content_type)) {
+    ngx_log_error(NGX_LOG_INFO, r->connection->log, 0,
+                  "RUM SDK injection skipped: not an HTML page");
+    return next_header_filter(r);
+  }
+
+  if (auto content_encoding = r->headers_out.content_encoding;
+      content_encoding != nullptr && content_encoding->value.len != 0) {
+    ngx_log_error(NGX_LOG_INFO, r->connection->log, 0,
+                  "RUM SDK injection skipped: compressed html content");
+    return next_header_filter(r);
+  }
+
+  state_ = state::searching;
+  injector_ = injector_create(cfg->rum_snippet);
+
+  // In case `Transfer-Encoding: chunk` is enabled no need to update the
+  // content length.
+  if (r->headers_out.content_length_n != -1) {
+    output_padding_ = true;
+    r->headers_out.content_length_n += cfg->rum_snippet->length;
+  }
+
+  // Set header now because it will be too late after
+  // TODO(@dmehala): write common function to insert HTTP Header
+  auto *h =
+      static_cast<ngx_table_elt_t *>(ngx_list_push(&r->headers_out.headers));
+  if (h == nullptr) {
+    state_ = state::error;
+    ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                  "RUM SDK injection failed: unable to add "
+                  "x-datadog-sdk-injected HTTP header");
+    return NGX_ERROR;
+  }
+
+  h->hash = 1;
+  ngx_str_set(&h->key, "x-datadog-sdk-injected");
```

Comment:
What is different in behavior beween `x-datadog-sdk-injected` and `x-datadog-rum-injected`? Should we be checking for the same headers and not use two?

Reply:
we should check only for `x-datadog-rum-injected`. The other one is from an older time where we didn't fully agree on the name.